# Assignment 1 The Power Plant Energy

## Project Description

The dataset contains 9568 data points collected from a Combined Cycle Power Plant over 6 years (2006-2011) when the plant was set to work with a full load. The dataset consists of hourly average ambient variables Temperature (T), Ambient Pressure (AP), Relative Humidity (RH), and Exhaust Vacuum (V) to predict the net hourly electrical energy output (EP) of the plant.

Features consist of hourly average ambient variables

- Temperature (T) in the range 1.81°C and 37.11°C,
- Ambient Pressure (AP) in the range 992.89-1033.30 millibar,
- Relative Humidity (RH) in the range 25.56% to 100.16%
- Exhaust Vacuum (V) in the range 25.36-81.56 cm Hg
- Net hourly electrical energy output (EP) 420.26-495.76 MW

The averages are taken from various sensors located around the plant that record the ambient variables every second. The variables are given without normalization.

Prepare a Python notebook (Jupyter NB) report, that contains a brief description of the project, and the features (variables); then split the dataset into train and test sets (80 & 20%), if necessary do the feature scaling, and build a model which can predict the EP (electrical energy output), using two methods and compare the results: Method 1. Multiple Regression and Method 2. SVM Regression.

## Importing Some Basic Libraries

In [121]:
import numpy as np 
import pandas as pd 
from matplotlib import pyplot as plt

## Importing the Dataset

In [122]:
dataset = pd.read_csv('Power Plant Data.csv')

## Showing the Dataset in a Table

In [123]:
dataset.head()

Ambient Temperature (C)   Exhaust Vacuum (cm Hg)  \
0                     14.96                   41.76   
1                     25.18                   62.96   
2                      5.11                   39.40   
3                     20.86                   57.32   
4                     10.82                   37.50   

   Ambient Pressure (milibar)  Relative Humidity (%)  \
0                     1024.07                  73.17   
1                     1020.04                  59.08   
2                     1012.16                  92.14   
3                     1010.24                  76.64   
4                     1009.23                  96.62   

   Hourly Electrical Energy output (MW)  
0                                463.26  
1                                444.37  
2                                488.56  
3                                446.48  
4                                473.90

## A Quick Review of the Data

In [124]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9568 entries, 0 to 9567
Data columns (total 5 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Ambient Temperature (C)               9568 non-null   float64
 1   Exhaust Vacuum (cm Hg)                9568 non-null   float64
 2   Ambient Pressure (milibar)            9568 non-null   float64
 3   Relative Humidity (%)                 9568 non-null   float64
 4   Hourly Electrical Energy output (MW)  9568 non-null   float64
dtypes: float64(5)
memory usage: 373.9 KB


 ## Separate the Input and Output

In [125]:
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

## Showing the Input Data in a Table Format

In [126]:
pd.DataFrame(X)

0      1        2      3
0     14.96  41.76  1024.07  73.17
1     25.18  62.96  1020.04  59.08
2      5.11  39.40  1012.16  92.14
3     20.86  57.32  1010.24  76.64
4     10.82  37.50  1009.23  96.62
...     ...    ...      ...    ...
9563  16.65  49.69  1014.01  91.00
9564  13.19  39.18  1023.67  66.78
9565  31.32  74.33  1012.92  36.48
9566  24.48  69.45  1013.86  62.39
9567  21.60  62.52  1017.23  67.87

[9568 rows x 4 columns]

## A Quick Check of the Output Data

In [127]:
pd.DataFrame(y)

0
0     463.26
1     444.37
2     488.56
3     446.48
4     473.90
...      ...
9563  460.03
9564  469.62
9565  429.57
9566  435.74
9567  453.28

[9568 rows x 1 columns]

## Splitting the Dataset into the Training Set and Test Set

In [128]:
from sklearn.model_selection import train_test_split

In [129]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [130]:
X_train.shape

(7654, 4)

In [131]:
X_test.shape

(1914, 4)

In [132]:
y_train.shape

(7654,)

# Method 1: Training the Multiple Linear Regression Model on the Training Set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

## Checking the Model with the Test set

In [133]:
# linear regression 
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
pd.DataFrame([y_pred,y_test])

0           1           2          3           4           5     \
0  457.255221  466.719274  440.366949  482.57801  474.880547  448.824227   
1  458.960000  463.290000  435.270000  484.31000  473.550000  456.300000   

         6           7           8           9     ...       1904        1905  \
0  440.149749  478.061406  466.985618  479.480139  ...  430.63655  428.319126   
1  436.020000  488.750000  469.750000  482.830000  ...  438.28000  432.190000   

         1906        1907        1908        1909        1910        1911  \
0  487.944282  464.437471  455.293903  445.676877  447.891673  476.405029   
1  494.870000  463.500000  449.330000  446.400000  457.120000  476.220000   

         1912        1913  
0  424.616097  463.911411  
1  440.290000  467.920000  

[2 rows x 1914 columns]

Check the accuracy of the multiple linear regression model.

In [134]:
from sklearn.metrics import r2_score

In [135]:
print(r2_score(y_test, y_pred))

0.9321860060402447


## Scaling the Features

In [136]:
# Since the input variables are all in a same range,we can skip this step
""""
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X[:, 3:6] = sc.fit_transform(X[:, 3:6])
pd.DataFrame(X)
"""""

'"\nfrom sklearn.preprocessing import StandardScaler\nsc = StandardScaler()\nX[:, 3:6] = sc.fit_transform(X[:, 3:6])\npd.DataFrame(X)\n'

## Measuring the Model Performance

In [137]:
from sklearn.metrics import mean_squared_error
RMSE=(mean_squared_error(y_pred,y_test))**0.5
print('RMSE: ', RMSE)

RMSE:  4.508879190536162


## Training and Testing Predictive Models

In [138]:
NewDataPt=np.array([15, 40, 1020, 70])
NewDataPt=NewDataPt.reshape(1, -1)
EP_Prediction= regressor.predict(NewDataPt)
print('The EP Prediction for the New Data Point = ', EP_Prediction)

The EP Prediction for the New Data Point =  [467.83988642]


# Method 2: Training the SVM Regression Model on the Training Set

In [139]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [140]:
y_train.shape

(7654,)

In [141]:
X_train.shape

(7654, 4)

### Feature Scaling

In [142]:
#need to do scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X_train = sc_X.fit_transform(X_train)
y_train = sc_y.fit_transform(y_train.reshape(-1, 1)) # we have to change the shape of y_train

### Training the Model

In [143]:
# Support vector machine
from sklearn.svm import SVR
SVRmodel = SVR(kernel = 'rbf')
SVRmodel.fit(X_train, y_train.reshape(len(y_train),))

SVR()

### Checking the SVM Model with the Test set

In [144]:
y_pred = SVRmodel.predict(sc_X.transform(X_test))
pd.DataFrame([y_pred,y_test])

0           1          2           3           4           5     \
0    0.154197    0.547249   -0.99024    1.833892    1.396393   -0.085902   
1  458.960000  463.290000  435.27000  484.310000  473.550000  456.300000   

        6           7           8           9     ...        1904        1905  \
0   -0.86267    1.607464    0.838121    1.762295  ...   -1.313182   -1.240862   
1  436.02000  488.750000  469.750000  482.830000  ...  438.280000  432.190000   

         1906        1907       1908        1909        1910        1911  \
0    2.023872    0.487897   -0.01347   -0.509876   -0.412424    1.291742   
1  494.870000  463.500000  449.33000  446.400000  457.120000  476.220000   

         1912        1913  
0   -1.261421    0.582211  
1  440.290000  467.920000  

[2 rows x 1914 columns]

In [145]:
# Let's make a few random check:
print("The Prediction Energy=",y_pred[103])
print("The Actual Energy=",y_test[103] )

The Prediction Energy= 0.20121586826818377
The Actual Energy= 458.61


In [146]:
NewDataPt=np.array([15, 40, 1020, 70])
NewDataPt=NewDataPt.reshape(1, -1)
EP_Prediction= SVRmodel.predict(NewDataPt)
print('The EP Prediction for the New Data Point = ', EP_Prediction)

The EP Prediction for the New Data Point =  [0.20753105]


### Evaluate Model Performance

Check the accuracy of the SVM regression model.

In [147]:
print(r2_score(y_test, y_pred))

-688.5474347973706


# Comparison between Multiple Regression and SVM Regression

We built two separate models in this notebook, one using multiple linear regression and the other using support vector machine regression. The accuracy of the multiple linear regression model is much higher at 93.2% compared to the SVM regression model at 38.4%.